### PR TITLE
Support semicolons as data table column separator (#1024)

### LINKED ISSUE #1024: pipe-characters in value of data table cell
I'd like to be able to do something like the following:

```groovy
        where:
            testee                                                | _
            restrictionChainElement                               | _
            restrictionChainElement & restrictionTwo.getClass()   | _
            (restrictionChainElement | restrictionTwo.getClass()) | _
            restrictionChainElement.negate()                      | _
```

where in the next to last line the pipe character is part of the actual cell value, calculating an _or_.

### PR DESCRIPTION
**Pull request by Vampire**
closes #1024

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1130)

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java [hunk 1]
```diff
@@ -381,6 +466,10 @@ private static void notAParameterization(ASTNode stat) throws InvalidSpecCompile
 "where-blocks may only contain parameterizations (e.g. 'salary << [1000, 5000, 9000]; salaryk = salary / 1000')");
   }
 
+  private static void dataTableHeaderMayOnlyContainVariableNames(ASTNode stat) throws InvalidSpecCompileException {
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: might be better to return the exception so that you can directly see that it exits in the code

`throw dataTableHeaderMayOnlyContainVariableNames(column.get(0));`

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy [hunk 2]
```diff
@@ -379,6 +546,16 @@ a | b | c
     4 | 5 || 9
   }
 
+  def "cells can be separated with any amount of semicolons"() {
+    expect:
+    a + b == c
+
+    where:
+    a ; b ;;   c
+    1 ; 2 ;;;  3
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: I don't know if we can fix this or if the compiler throws it away already, but I don't really like this. IMHO every row should have the same identifiers

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy [hunk 3]
```diff
@@ -166,6 +247,29 @@ a | a
...
+    [a, b, c, d] == [1, 2, 3, 4]
+
+    where:
+    a << [1]
+    b ; c
+    2 ; 3
+    d = c + 1
+  }
+
+  def "tables with different separators can be mixed"() {
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: I'm missing a test that describes your original problem with using `|` in a data table value

File: spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    Math.max(a, b) == c
+    Math.max(d, e) == f
+
+    where:
+    a | b || c
+    1 | 3 || 3
+    7 | 4 || 7
+    0 | 0 || 0
+
+    ; d ; e ;; f ;
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: why are there `;` before and after the row

File: spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java (hunk 1 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java]: Sure, I totally agree, I just followed the tactic mandated for by `notAParameterization`.
I changed both now.

File: spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy (hunk 2 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/DataTables.groovy]: Well, we also do not enforce this for pipe and double-pipe and there are even tests that make sure you can freely mix thme.
But anyway, as you assumed, the parser throws them out already and there is no way to differentiate this case as far as I have seen.

File: spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy (hunk 4 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: Just to showcase that you can.
This way you can visually add a left and right border so that it looks more table-like.
If my data table separator PR is merged you can even add an upper border and with a bit distance bottom border like
```groovy
    ______________
    ; d ; e ;; f ;
    ; 1 ; 3 ;; 3 ;
    ; 7 ; 4 ;; 7 ;
    ; 0 ; 0 ;; 0 ;
    ______________
```

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: I think it is better to introduce one concept at a time and would move that into a separate example. In this case I think it makes more sense to have the same layout as the pipe table above, not every new user will directly see what is necessary and what is not.

Documentation is also treated as a kind of a style guide, and while it is possible to encase your table that way and we may mention it, I would not recommend it. Groovy is often referred to as java without semicolons, so having more visual noise is often frowned upon.

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: Well, the example right before shows it without the borders:
grafik
But I can of course add this as an additional example.

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: Another reason why I combined them was, because the first semicolon in the first line must have a preceding statement, because you cannot end the label with a semicolon.
This is **not** valid Groovy syntax:
```groovy
    where:
    ; a ; b ;; c ;
    ; 1 ; 3 ;; 3 ;
    ; 7 ; 4 ;; 7 ;
    ; 0 ; 0 ;; 0 ;
```
With the other table in front it was valid.
So don't wonder that the new example will be
```groovy
    where:
    ____________=_
    ; a ; b ;; c ;
    ; 1 ; 3 ;; 3 ;
    ; 7 ; 4 ;; 7 ;
    ; 0 ; 0 ;; 0 ;
```
If the data table separator PR gets merged, the `=` can be replaced by an `_`, then it looks properly like
```groovy
    where:
    ______________
    ; a ; b ;; c ;
    ; 1 ; 3 ;; 3 ;
    ; 7 ; 4 ;; 7 ;
    ; 0 ; 0 ;; 0 ;
```

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: I get more and more convinced to **not** include those examples at all, if they have all these edge cases. You can still use them in your projects if you find them helpful, but I don't think it is good to add them to the official documentation. We should present them to be a drop in replacement to the `|`/`||` style nothing more nothing less.

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: "All these edge cases"?
Maybe I'm too tunnel-focused right now, but I see only one edge case, that the `where:` label must not be followed by a semicolon directly.
As the data table serparator PR was already merged, I now also adapted the example here to be the proper one without equals sign.
I rebased this PR and fixed the conflicts.
Maybe you want to have a look again now and decide then.
If you really want the example removed, I'll remove it, I just don't get what you mean with "All these edge cases".

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: Well the other is that you can't do the same with `|`. I prefer to remove them from the documentation.

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: You can:
```groovy
    where:
    ______________
    ; a | b || c ;
    ; 1 | 3 || 3 ;
    ; 7 | 4 || 7 ;
    ; 0 | 0 || 0 ;
```
(I know this is not what you meant :-) )

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/docs/datadriven/v7/MathSpec.groovy]: done
